### PR TITLE
Sasha/configure

### LINKED ISSUE #5559: Improve teleport configure subcommand
### What

It should ask you a couple of questions and generate a valid teleport.yaml and write it to the disk.

* What is your cluster name? (For example example.com). Cluster name can not be changed, so pick a unique name. If you choose a name that matches your domain name, configuration will be easier.

* Backend? Dir by default.

* Acme. Should offer acme for auto cert generation

* Should print the survey at the end.

## Test

`teleport configure --test /path` should test config for errors.

### PR DESCRIPTION
**Pull request by klizhentas**
Fixes #5559

Configure with -o file create file /etc/teleport.yaml.
This commit optimizes configure for getting users started instead of generating sample
files.

```bash
teleport configure -o file --cluster-name=example.com --acme --acme-email=alice@example.com
```

### CODE REVIEW

File: tool/teleport/common/teleport.go [hunk 1]
```diff
@@ -165,6 +172,17 @@ func Run(options Options) (executedCommand string, conf *service.Config) {
...
+	// dump flags
+	dump.Flag("cluster-name",
+		"Unique cluster name, e.g. example.com.").StringVar(&dumpFlags.ClusterName)
+	dump.Flag("output",
+		"Write to stdout, default config file or custom path").Short('o').Default(
+		teleport.SchemeStdout + "://").StringVar(&dumpFlags.output)
+	dump.Flag("acme",
+		"Get automatice certificate from Letsencrypt.org using ACME.").BoolVar(&dumpFlags.ACMEEnabled)
+	dump.Flag("acme-email",
+		"Email to receive update from Letsencrypt.org.").StringVar(&dumpFlags.ACMEEmail)
```
Comment (webvictim) [inline tool/teleport/common/teleport.go:184]:
```suggestion
		"Email to receive updates from Letsencrypt.org.").StringVar(&dumpFlags.ACMEEmail)
```

**Comment by klizhentas** [comment]
@russjones @r0mant I think this is ready to go, I'd like to get this in 6.0 so I can fix our getting started guide's scary parts.

File: tool/teleport/common/teleport.go [hunk 2]
```diff
@@ -165,6 +172,17 @@ func Run(options Options) (executedCommand string, conf *service.Config) {
 	scpc.Flag("local-addr", "local address which accepted the request").StringVar(&scpFlags.LocalAddr)
 	scpc.Arg("target", "").StringsVar(&scpFlags.Target)
 
+	// dump flags
+	dump.Flag("cluster-name",
+		"Unique cluster name, e.g. example.com.").StringVar(&dumpFlags.ClusterName)
+	dump.Flag("output",
+		"Write to stdout, default config file or custom path").Short('o').Default(
```
Comment (r0mant) [inline tool/teleport/common/teleport.go:179]: Would be nice to give a couple examples of how to write to stdout, default file or provide a custom path. It's not really clear just reading the message.

File: tool/teleport/common/teleport.go [hunk 3]
```diff
@@ -238,12 +256,48 @@ func onStatus() error {
...
 	}
-	fmt.Printf("%s\n%s\n", sampleConfComment, sfc.DebugDumpToYAML())
+	switch uri.Scheme {
+	case teleport.SchemeStdout:
+		fmt.Printf("%s\n%s\n", sampleConfComment, sfc.DebugDumpToYAML())
+	case teleport.SchemeFile, "":
+		if uri.Path == "" {
+			return trace.BadParameter("missing path in --output=%q", uri)
+		}
+		f, err := os.OpenFile(uri.Path, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0600)
```
Comment (r0mant) [inline tool/teleport/common/teleport.go:280]: There's `teleport.FileMaskOwnerOnly` constant for `0600`.

File: tool/teleport/common/teleport.go [hunk 4]
```diff
@@ -238,12 +256,48 @@ func onStatus() error {
...
+		err = trace.ConvertSystemError(err)
+		if err != nil {
+			if trace.IsAlreadyExists(err) {
+				return trace.AlreadyExists("will not overwrite existing file %v, rm -f %v and try again", uri.Path, uri.Path)
+			}
+			return trace.Wrap(trace.ConvertSystemError(err), "could not write to config file, missing sudo?")
```
Comment (r0mant) [inline tool/teleport/common/teleport.go:286]: Didn't you already convert `err` to system error above?

File: tool/teleport/common/teleport.go [hunk 5]
```diff
@@ -64,11 +70,12 @@ func Run(options Options) (executedCommand string, conf *service.Config) {
 	// define global flags:
 	var ccf config.CommandLineFlags
 	var scpFlags scp.Flags
+	var dumpFlags dumpFlags
 
 	// define commands:
 	start := app.Command("start", "Starts the Teleport service.")
 	status := app.Command("status", "Print the status of the current SSH session.")
-	dump := app.Command("configure", "Print the sample config file into stdout.")
+	dump := app.Command("configure", "Generate config file.")
```
Comment (r0mant) [inline tool/teleport/common/teleport.go:78]: I would expand this a bit to explain what will be generated e.g. something like `Generate config file that will start Teleport auth, proxy and SSH services`.

File: tool/teleport/common/teleport.go [hunk 6]
```diff
@@ -165,6 +172,17 @@ func Run(options Options) (executedCommand string, conf *service.Config) {
...
+	dump.Flag("cluster-name",
+		"Unique cluster name, e.g. example.com.").StringVar(&dumpFlags.ClusterName)
+	dump.Flag("output",
+		"Write to stdout, default config file or custom path").Short('o').Default(
+		teleport.SchemeStdout + "://").StringVar(&dumpFlags.output)
+	dump.Flag("acme",
+		"Get automatic certificate from Letsencrypt.org using ACME.").BoolVar(&dumpFlags.ACMEEnabled)
+	dump.Flag("acme-email",
+		"Email to receive updates from Letsencrypt.org.").StringVar(&dumpFlags.ACMEEmail)
+
```
Comment (r0mant) [inline tool/teleport/common/teleport.go:185]: It looks like `SampleFlags` also have `AddLicense` field but there's no corresponding CLI flag. Do you need to add it as well?

**Comment by klizhentas** [comment]
@r0mant @russjones please take another look, thanks!

**Review by r0mant — APPROVED** [review]

File: tool/teleport/common/teleport.go [hunk 7]
```diff
@@ -237,13 +251,74 @@ func onStatus() error {
...
-	sfc, err := config.MakeSampleFileConfig()
+func onConfigDump(flags dumpFlags) error {
+	if err := flags.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	uri, err := url.Parse(flags.output)
+	if err != nil {
+		return trace.BadParameter("could not parse output value %q, use --output=%q",
+			flags.output, defaults.ConfigFilePath)
+	}
```
Comment (russjones) [inline tool/teleport/common/teleport.go:277]: Minor: Pass through https://golang.org/pkg/path/filepath/#EvalSymlinks to resolve symlinks for the user.

**Review by russjones — APPROVED** [review]

**Pull request merged**